api/src/verify_token.rs

```diff
@@ -0,0 +1,222 @@
+        },
+        Err(_) => Ok(PossibleToken::Unverified),
+    }
+}
+
+pub async fn create(
+    db: &impl Database,
+    id: &str,
+    expires_at: DateTime<Utc>,
+) -> Result<UnverifiedToken>
+{
+    if let Ok(bytes) = db.get(&id).await {
+        if serde_json::from_slice::<VerifiedToken>(&bytes).is_ok() {
+            return Err(VerifyTokenError::Done);
+        }
+
+        match serde_json::from_slice::<UnverifiedToken>(&bytes) {
+            Ok(_) => {},
+            Err(e) => {
+                return Err(VerifyTokenError::Serde { source: e });
+            },
+        };
+
+        // otherwise overwrite the unverified token, which corresponds
+        // to the case of "resend the confirmation email", etc.
+    }
+
+    let secret = Mu::new();
+    let encoded_secret = base64::encode(secret.0);
+
+    let t = UnverifiedToken {
+        schema_version: 0,
+        secret: encoded_secret,
+        expires_at: expires_at,
+    };
+
+    let bytes = match serde_json::to_vec(&t) {
+        Ok(b) => b,
+        Err(e) => return Err(VerifyTokenError::Serde { source: e }),
+    };
+
+    match db.put(id, &bytes).await {
+        Ok(_) => Ok(t),
+        Err(_) => Err(VerifyTokenError::Unknown),
+    }
+}
+
+pub async fn verify(
+    db: &impl Database,
+    id: &str,
+    secret: &str,
+    method: VerifyMethod,
+) -> Result<VerifiedToken>
+{
+    if let Ok(bytes) = db.get(id).await {
+        if serde_json::from_slice::<VerifiedToken>(&bytes).is_ok() {
+            return Err(VerifyTokenError::Done);
+        }
+
+        match serde_json::from_slice::<UnverifiedToken>(&bytes) {
+            Ok(u) => match u.schema_version {
+                0 => {
+                    if Utc::now() > u.expires_at {
+                        return Err(VerifyTokenError::Expired);
+                    }
+
+                    if u.secret != secret {
+                        return Err(VerifyTokenError::Secret);
+                    }
+
+                    let v = VerifiedToken { schema_version: 0, method: method };
+
+                    let bytes = match serde_json::to_vec(&v) {
+                        Ok(b) => b,
+                        Err(e) => {
+                            return Err(VerifyTokenError::Serde { source: e });
+                        },
+                    };
+
+                    return match db.put(id, &bytes).await {
+                        Ok(_) => Ok(v),
+                        Err(_) => Err(VerifyTokenError::Unknown),
+                    };
+                },
+                _ => {
+                    return Err(VerifyTokenError::Schema);
+                },
+            },
+            Err(e) => {
+                return Err(VerifyTokenError::Serde { source: e });
+            },
+        };
+    }
+
+    Err(VerifyTokenError::NotFound)
+}
+
+#[cfg(test)]
+mod tests
+{
+    use super::*;
+    use chrono::Duration;
+
+    #[cfg(not(feature = "s3"))]
+    use crate::store::FileStore;
+
+    #[cfg(not(feature = "s3"))]
```

Comment:
why can't these run when using the s3 backend?

Reply:
We can but i don't really want to hit s3 in the tests. We can set up minio in the test environment if you want.